cloud/amazon/sts_assume_role.py

```diff
@@ -0,0 +1,166 @@
+'''
+
+EXAMPLES = '''
+# Basic example of assuming a role
+tasks:
+- name: assume a role in account 123456789012
+  sts_assume_role: role_arn="arn:aws:iam::123456789012:role/someRole" session_name="someRoleSession"
+
+- name: display temporary credentials
+  debug: "temporary credentials for the assumed role are {{ ansible_temp_credentials }}"
+  
+- name: use temporary credentials for tagging an instance in account 123456789012 
+  ec2_tag: resource=i-xyzxyz01 region=us-west-1 state=present
+    args:
+      aws_access_key: "{{ ansible_temp_credentials.access_key }}"
+      aws_secret_key: "{{ ansible_temp_credentials.secret_key }}"
+      security_token: "{{ ansible_temp_credentials.session_token }}"
+    
+      tags:
+        Test: value
+'''
+
+import sys
+import time
+
+try:
+    import boto.sts
+    
+except ImportError:
+    print "failed=True msg='boto required for this module'"
+    sys.exit(1)
+
+def sts_connect(module):
+
+    """ Return an STS connection"""
+
+    region, ec2_url, boto_params = get_aws_connection_info(module)
+
+    # If we have a region specified, connect to its endpoint.
+    if region:
+        try:
+            sts = connect_to_aws(boto.sts, region, **boto_params)
+        except boto.exception.NoAuthHandlerFound, e:
+            module.fail_json(msg=str(e))
+    # Otherwise, no region so we fallback to connect_sts method
+    else:
+        try:
+            sts = boto.connect_sts(**boto_params)
+        except boto.exception.NoAuthHandlerFound, e:
+            module.fail_json(msg=str(e))
+
+
+    return sts
+
+def assumeRole():
+    data = sts.assume_role()
+    return data
+
+def main():
+    argument_spec = ec2_argument_spec()
+    argument_spec.update(dict(
+            role_arn = dict(required=True),
+            role_session_name = dict(required=True),
+            duraction_seconds = dict(),
+            external_id = dict(),
+            policy = dict(),
+            mfa_serial_number = dict(),
+            mfa_token = dict(),
+        )
+    )
+    module = AnsibleModule(argument_spec=argument_spec)
+
+    role_arn = module.params.get('role_arn')
+    role_session_name = module.params.get('role_session_name')
+    policy = module.params.get('policy')
+    duraction_seconds = module.params.get('duraction_seconds')
+    external_id = module.params.get('external_id')
+    mfa_serial_number = module.params.get('mfa_serial_number')
+    mfa_token = module.params.get('mfa_token')
+  
+    sts = sts_connect(module)
+    
+    temp_credentials = {}
+    
+    try:
+        temp_credentials = sts.assume_role(role_arn, role_session_name, policy, duraction_seconds,
+                                           external_id, mfa_serial_number, mfa_token).credentials.__dict__
+    except boto.exception.BotoServerError, e:
+        module.fail_json(msg='Unable to assume role {0}, error: {1}'.format(role_arn, e))
+    result = dict(changed=False, ansible_facts=dict(ansible_temp_credentials=temp_credentials))

```

Comment:
Do we really want to be returning changed=False here?  I would think that generating a temporary security credential is a big change.

I also don't like the idea of setting the credentials as `ansible_temp_credentials` within ansible_facts. I think you should rely on the user registering the result instead.  Since STS can be used for cross-account access I can see a potential case where a playbook needs to interact with multiple AWS accounts, so I could see a user wanting to do something like this:

    - name: assume a role in account 123456789012
       sts_assume_role: role_arn="arn:aws:iam::123456789012:role/someRole" session_name="someRoleSession"
       register: acct1

    - name: assume a role in account 987654321000
        sts_assume_role: role_arn="arn:aws:iam::987654321000:role/someRole" session_name="someRoleSession"
        register: acct2

    - name: delete bucket in account1
        s3: bucket=mybucket mode=delete aws_access_key= "{{ acct1.access_key }}" ...

    - name: delete bucket in account2
        s3: bucket=mybucket mode=delete aws_access_key= "{{ acct2.access_key }}" ...

The name `ansible_temp_credentials` itself also bugs me, especially as used in `ansible_facts`.  If it's left as part of `ansible_facts` then I'd suggest renaming it to something more descriptive like `ansible_aws_temporary_credentials`.  `ansible_temporary_credentials` implies they're related to ansible in general, when in fact they're specific and unique to the various AWS modules.  However, as I mentioned above I think it would be better to remove this from `ansible_facts`, so I'd suggest something like:

    result =  dict(changed=True, sts=temp_credentials)

Which should allow users to reference the credentials this way:

    - name: delete bucket in account1
        s3: bucket=mybucket mode=delete aws_access_key= "{{ acct1.sts.access_key }}" ...

Other than these comments and the notes made by @rickmendes I think this module looks good.